### PR TITLE
Mdolotova/grok 12459/handling malformed data

### PR DESCRIPTION
**Pull request by MariaDolotova**

### CODE REVIEW

File: packages/Chem/src/chem-searches.ts [hunk 1]
```diff
@@ -55,18 +55,22 @@ function _chemFindSimilar(molStringsColumn: DG.Column, fingerprints: BitArray[],
 // Only this function receives {sorted} in settings
 function _chemGetSimilarities(queryMolString: string, fingerprints: BitArray[]): number[] {
   const distances = new Array(fingerprints.length).fill(0.0);
-  const sample = chemGetFingerprint(queryMolString, Fingerprint.Morgan);
-  for (let i = 0; i < fingerprints.length; ++i)
-    distances[i] = tanimotoSimilarity(fingerprints[i], sample);
+  try{
```
Comment (LaykDimon) [inline packages/Chem/src/chem-searches.ts:58]: space before brace

File: packages/Chem/src/utils/malformed-data-utils.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+import * as ui from 'datagrok-api/ui';
+
+export function malformedDataWarning(fingerprintCol: BitArray[], dataframe: DG.DataFrame): (number | null)[] {
+    const malformedData = fingerprintCol.map((it: BitArray, idx: number) => it.allFalse ? idx : null).filter((it) => it !== null);
+      if (malformedData.length) {
+        const message = ui.divText(`${malformedData.length} molecules with indexes: ${malformedData.map((it: number | null) => it! + 1)
+          .join(',')} are empty or possibly malformed and are not included in analysis`);
+        const selectRowsButton = ui.button('Select', ()=> {
+          for (const i of malformedData)
+          dataframe.selection.set(i!, true);
```
Comment (LaykDimon) [inline packages/Chem/src/utils/malformed-data-utils.ts:13]: tabulation before dataframe

File: libraries/ml/src/viewers/activity-cliffs.ts [hunk 3]
```diff
@@ -84,26 +84,21 @@ export async function getActivityCliffs(df: DG.DataFrame, seqCol: DG.Column, enc
...
     embedAxesNames: axesNames,
     options: seqSpaceOptions,
   };
 
   const { distance, coordinates } = await seqSpaceFunc(seqSpaceParams);
-  const coordsWithInsertedEmptyVals = getEmbeddingColsWithInsertedEmptyVals(coordinates, emptyValsIdxs, df.rowCount);
-  for (const col of coordsWithInsertedEmptyVals) {
+  for (const col of coordinates) {
     df.columns.add(col);
   }
```
Comment (vdyma) [inline libraries/ml/src/viewers/activity-cliffs.ts:98]: Here we could get rid of curly braces

File: packages/Chem/src/analysis/chem-space.ts [hunk 4]
```diff
@@ -9,19 +9,33 @@ import {Fingerprint} from '../utils/chem-common';
...
+    }
+  })
   const cols: DG.Column[] = spaceParams.embedAxesNames.map((name: string, index: number) => DG.Column.fromFloat32Array(name, chemSpaceResult.embedding[index]));
   return {distance: chemSpaceResult.distance, coordinates: new DG.ColumnList(cols)};
 }
 
+function setNullForMalformedData(matrix: Matrix, idx: number) {
+  for (const col of matrix) {
+    col[idx] = DG.FLOAT_NULL;
+  };
```
Comment (vdyma) [inline packages/Chem/src/analysis/chem-space.ts:36]: Can get rid of curly braces and never set semicolon after closing curly braces

**Pull request merged**